File: app/src/main/java/com/techipinfotech/onlinestudy1/HomeApi.java
package com.techipinfotech.onlinestudy1;

import android.util.Log;

import java.util.List;
import java.util.concurrent.TimeUnit;

import com.techipinfotech.onlinestudy1.model.ChaptersItem;
import com.techipinfotech.onlinestudy1.model.JSONResponse;
import com.techipinfotech.onlinestudy1.model.Received;
import okhttp3.OkHttpClient;
import okhttp3.ResponseBody;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Call;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;
import retrofit2.http.GET;
import retrofit2.http.Path;
import retrofit2.http.Streaming;
import retrofit2.http.Url;

public class HomeApi {

    public static final String base_url = API.HOME_URL.toString();

    private static ApiService apiService = null;

    public static ApiService getApiService() {


        if (apiService == null) {
            //create Ok HttP Client
            OkHttpClient.Builder builder = new OkHttpClient.Builder();
            HttpLoggingInterceptor httpLoggingInterceptor = new HttpLoggingInterceptor(s -> Log.d("ASA", s));
            httpLoggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
            builder.addInterceptor(httpLoggingInterceptor)
                    .readTimeout(60, TimeUnit.SECONDS)
                    .connectTimeout(60, TimeUnit.SECONDS);
//            create
            Retrofit retrofit = new Retrofit.Builder()
                    .baseUrl(base_url)
                    .addConverterFactory(GsonConverterFactory.create())
                    .client(builder.build())
                    .build();

            apiService = retrofit.create(ApiService.class);
        }
        return apiService;

    }

    public interface ApiService {

        @GET("getjsondata/{mobile_no}")
        Call<List<JSONResponse>> getjsondata(@Path("mobile_no") String mobile_no);

        @GET("getsubject_jsondata/{student_id}/{subject_id}")
        Call<List<ChaptersItem>> getSubjectJsonData(@Path("student_id") String student_id,
                                                    @Path("subject_id") String subject_id);

        @GET("updatevideoviewedstatus/{username}/{material_id}")
        Call<Received> updateViewedStatus(
                @Path("username") String username,
                @Path("material_id") String material_id
        );

        @Streaming
        @GET
        Call<ResponseBody> downloadPDF(@Url String fileUrl);


    }

}

File: app/src/main/java/com/techipinfotech/onlinestudy1/HomeApi.kt
package com.techipinfotech.onlinestudy1

import android.util.Log
import com.techipinfotech.onlinestudy1.model.ChaptersItem
import com.techipinfotech.onlinestudy1.model.ContentItem
import com.techipinfotech.onlinestudy1.model.JSONResponse
import com.techipinfotech.onlinestudy1.model.Received
import com.techipinfotech.onlinestudy1.model.TopicsItem
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Call
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.GET
import retrofit2.http.Path
import java.util.concurrent.TimeUnit

object HomeApi {

    private val BASE_URL = API.HOME_URL.toString()

    private var apiService: ApiService? = null

    fun getApiService(): ApiService {
        if (apiService == null) {
            val httpLoggingInterceptor = HttpLoggingInterceptor { message -> Log.d("ASA", message) }
            httpLoggingInterceptor.level = HttpLoggingInterceptor.Level.BODY

            val okHttpClient = OkHttpClient.Builder()
                .addInterceptor(httpLoggingInterceptor)
                .readTimeout(60, TimeUnit.SECONDS)
                .connectTimeout(60, TimeUnit.SECONDS)
                .build()

            val retrofit = Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .client(okHttpClient)
                .build()

            apiService = retrofit.create(ApiService::class.java)
        }
        return apiService!!
    }

    interface ApiService {

        @GET("getjsondata/{mobile_no}")
        fun getJsonData(@Path("mobile_no") mobileNo: String): Call<List<JSONResponse>>

//        @GET("getsubject_jsondata/{student_id}/{subject_id}")
//        fun getSubjectJsonData(
//            @Path("student_id") studentId: String,
//            @Path("subject_id") subjectId: String
//        ): Call<List<ChaptersItem>>

        @GET("getchapter/{subject_id}/{student_id}")
        fun getChapterJsonData(
            @Path("subject_id") subjectId: String,
            @Path("student_id") studentId: String
        ): Call<List<ChaptersItem>>

        @GET("gettopic/{subject_id}/{chapter_id}/{student_id}")
        fun getTopicJsonData(
            @Path("subject_id") subjectId: String,
            @Path("chapter_id") chapterId: String,
            @Path("student_id") studentId: String
        ): Call<List<TopicsItem>>

        @GET("getvideo/{subject_id}/{chapter_id}/{topic_id}/{student_id}")
        fun getVideo(
            @Path("subject_id") subjectId: String,
            @Path("chapter_id") chapterId: String,
            @Path("topic_id") topicId: String,
            @Path("student_id") studentId: String,
        ): Call<List<ContentItem>>

        @GET("updatevideoviewedstatus/{username}/{material_id}")
        fun updateViewedStatus(
            @Path("username") username: String,
            @Path("material_id") materialId: String
        ): Call<Received>

    }
}
